.github/workflows/build.yml

```diff
@@ -0,0 +1,171 @@
+name: build
+
+on:
+  push:
+    tags:
+    - v*
+    branches:
+    - master
+  pull_request:
+    branches:
+    - master
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.head_ref }}
+  cancel-in-progress: true
+
+env:
+  STABLE_PYTHON_VERSION: "3.11"
+  CIBW_BUILD_FRONTEND: build
+  CIBW_TEST_COMMAND: >
+    python -c
+    "from pact import EachLike;
+    assert EachLike(1).generate() == {'json_class': 'Pact::ArrayLike', 'contents': 1, 'min': 1}
+    "
+
+jobs:
+  build-x86_64:
+    name: Build wheels on ${{ matrix.os }} (x86, 64-bit)
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+        - os: ubuntu-latest
+          archs: x86_64
+        - os: macos-latest
+          archs: x86_64
+        - os: windows-latest
+          archs: AMD64
+
+    steps:
+    - uses: actions/checkout@v4
+      with:
+        # Fetch all tags
+        fetch-depth: 0
+
+    - name: Create wheels
+      uses: pypa/cibuildwheel@v2.15.0
+      env:
+        CIBW_ARCHS: ${{ matrix.archs }}
+
+    - name: Upload wheels
+      uses: actions/upload-artifact@v3
+      with:
+        name: wheels
+        path: ./wheelhouse/*.whl
+        if-no-files-found: error
+
+  build-x86:
+    name: Build wheels on ${{ matrix.os }} (x86, 32-bit)
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+        - os: windows-latest
+          archs: x86
+
+    steps:
+    - uses: actions/checkout@v4
+      with:
+        # Fetch all tags
+        fetch-depth: 0
+
+    - name: Create wheels
+      uses: pypa/cibuildwheel@v2.15.0
+      env:
+        CIBW_ARCHS: ${{ matrix.archs }}
+
+    - name: Upload wheels
+      uses: actions/upload-artifact@v3
+      with:
+        name: wheels
+        path: ./wheelhouse/*.whl
+        if-no-files-found: error
+
+  build-arm64:
+    name: Build wheels on ${{ matrix.os }} (arm64)
+    runs-on: ${{ matrix.os }}
+    # As this requires emulation, it's not worth running on PRs
+    if: >-
+      github.event_name == 'push' &&
+      startsWith(github.event.ref, 'refs/tags')
```

Comment:
This has been tested on GitHub Actions and is confirmed to work, along with the following `check` step.

Reply:
How slow is slow? ahh just seen 40 ish mins

<img width="327" alt="Screenshot 2023-09-14 at 13 18 16" src="https://github.com/pact-foundation/pact-python/assets/19932401/33764014-7185-4243-bb1e-570ce2c00aa5">

Reply:
In terms of the computation, you are correct, nothing is _too_ computationally complex.

I make use of [cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/) to manage the building of wheels against the various targets. It does handle a couple of extra things on top of just building the wheel (it can run the test suite, or in the above case, just tests a rather trivial use of Pact), it audits the wheel, and for binary wheels, also ensures that all necessary system libraries are packaged as well. It also repeats this for all main versions of Python.

I suspect that the GitHub Runners are quite highly optimised for running just Linux, and when you have a containerised environment with `qemu` handling the architecture virtualisation, things evidently slow down quite drastically.

The 50 minute runtime ultimately is not _awful_, but I certainly wouldn't want to force every PR to wait on this to succeed.

Reply:
40/50 mins is similar to what I see using cross (qemu) and docker building ruby from source, cross arch.

I'm surprised to see it take that long, maybe that is because the audit and stuff also running under emulation.

Anyhoo, as we won't build these that often its probably of no consequence

```console
🕙15:32:53 [🧱 INT] ❯du -sh dist/*
 18M    dist/pact_python-2.0.1-cp311-cp311-macosx_10_16_arm64.whl
 60K    dist/pact_python-2.0.1.tar.gz
 ```
 
File sizes are much nicer now with the indiv packaged wheels. But I assume the pure python stuff in the smaller package, and the pact-ruby-standalone is embedded in the whl. I still can't fathom why it takes so long to do python arm64 things.

anyway, its other peoples machines (ie github runners) so its fine!